### PR TITLE
Speed up `enry.IsVendor`

### PR DESCRIPTION
**Pull request by zeripath**
`enry.IsVendor` is kinda slow as it simply iterates across all regexps.
This PR ajdusts the regexps to combine them to make this process a
little quicker.

Related #15143

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by silverwind** [comment]
Seems like a good idea. [Linguist does similar](https://github.com/github/linguist/blob/841803f28777daa3442aa0fab8ca6cd16d674842/lib/linguist/blob_helper.rb#L220), but they simply `|` the individual regex parts. Is this not an option here? Also, should upstream this.

**Comment by zeripath** [comment]
* Simply `|` the individual parts leads to a 4 times slow down.
* Sorting the regexps and `|` is 3 times slower.
* making the groups non capturing makes it twice as slow
* Placing all of the `^` first and making it common to the ones that use it is around equal
* Following those by `(^|/)` and with the common prefix extracted is 2 times faster. 

This iterative process should suggest that some further simplifications should speed things up further. However if we drop full regexp compatibility we should be able to make it faster with a path trie. (I recently came across a library that looked reasonable for this - but its name escapes me.)

**Comment by zeripath** [comment]
My suspicion is that go's regexp simplifier is just not as good as ruby's or even that GitHub is using something like hyperscan to make it faster.

It's really worth noting that twice as fast here is still not really enough - it really would be better to be of the order of at least 10 times as fast. 

Now in the associated issue the profile indicated more than a few other places that needed speed ups - however I suspect that some of these may be already on their way to being improved. (For example I think since the chi migration we shouldn't be listing branches or tags to do a http push anymore and we don't do a hash to assert if a password is set - we do however test password authentication twice during pushing!!)

**Comment by silverwind** [comment]
If `enry.isVendor` is a pure function (e.g. not reading files like `.gitattributes`), memoization could be another big gain if it's being called repeatedly on the same paths.

**Comment by zeripath** [comment]
I don't think we can assume that but I guess there are likely to be some common paths that could be captured. It's possible though that a trie can be just as fast as a cache.

Certainly though there are lots of places we could be memoizing by throwing stuff in to caches or precomputing - in particular commit verification.

**Comment by lafriks** [comment]
Maybe this could be contributed upstream?

File: modules/analyze/vendor.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	sb.WriteString(")")
+	combined := sb.String()
+	isVendorRegExp = regexp.MustCompile(combined)
+}
+
+// IsVendor returns whether or not path is a vendor path.
+func IsVendor(path string) bool {
+	return isVendorRegExp.MatchString(path)
+}
```
Comment (lunny) [inline modules/analyze/vendor.go:70]: Could we have some unit tests here?

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Should definitely upstream this, including the tests which already seem more extensive than [enry's own](https://github.com/go-enry/go-enry/blob/d7f6b27b7dacf7af7a37daa11e3c23be1d1eb34f/utils_test.go#L14-L35). Thought I guess as a short-term solution it's fine.

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Pull request merged**